mpegts/mpegts_demuxer.cpp

```diff
@@ -1,210 +1,309 @@
-                        lEsFrame.mCompleted = true;
-                        esOutCallback(lEsFrame);
-                    }
-                    lEsFrame.reset();
+            mStreamPidMap[mInfo->mStreamType] = mInfo->mElementaryPid;
+        }
+        mPmtIsValid = true;
+
+        if (streamInfoCallback != nullptr) {
+            mPmtHeader.print(LogLevel::kTrace, streamInfoCallback);
+        }
+    }
+
+    return mPmtIsValid;
+}
+
+bool MpegTsDemuxer::parsePes(const mpegts::TsHeader& rTsHeader, uint8_t randomAccessIndicator, mpegts::SimpleBuffer& rSb, size_t payloadSize) {
+    size_t lPesStartPos = rSb.pos();
+
+    EsFrame& lEsFrame = mEsFrames[rTsHeader.mPid];
+    if (rTsHeader.mPayloadUnitStartIndicator == 0x01) {
+        // TODO: Transport streams with no PES packet size set on video streams will not output the very
+        // last frame, since there will be no final packet with mPayloadUnitStartIndicator set. Add a flush
+        // function to the demuxer API that the user can call once it has ran out of input data.
+
+        if (!lEsFrame.mData->empty()) {
+            // We have an old packet with data, flush and reset it before starting a new one
+            if (esOutCallback) {
+                if (lEsFrame.mExpectedPesPacketLength == 0) {
+                    lEsFrame.mCompleted = true;
+                } else {
+                    // Expected length is set, but we didn't get enough data, deliver what we have as broken
+                    lEsFrame.mBroken = true;
                 }
+                esOutCallback(lEsFrame);
             }
-        } else if (mPcrId != 0 && mPcrId == lTsHeader.mPid) {
-            AdaptationFieldHeader lAdaptField;
-            lAdaptField.decode(rIn);
-            lDiscontinuityIndicator = lAdaptField.mDiscontinuityIndicator;
-            uint64_t lPcr = readPcr(rIn);
-            if (pcrOutCallback) {
-                pcrOutCallback(lPcr);
-            }
-            checkContinuityCounter(lTsHeader, lDiscontinuityIndicator);
+            lEsFrame.reset();
         }
-        rIn.skip(188 - (rIn.pos() - lPos));
+
+        // Potential old package has been flushed, set random access indicator and start new package
+        lEsFrame.mRandomAccess = randomAccessIndicator;
+
+        PESHeader lPesHeader;
+        lPesHeader.decode(rSb);
+        if (lPesHeader.mMarkerBits != 0x02) {
+            LOG(LogLevel::kWarning, "Corrupt PES header, marker bits not as expected");
+            return false;
+        }
+
+        const size_t lPositionAfterPesHeaderDataLength = rSb.pos();
+
+        lEsFrame.mStreamId = lPesHeader.mStreamId;
+        lEsFrame.mExpectedPesPacketLength = lPesHeader.mPesPacketLength;
+        if (lPesHeader.mPtsDtsFlags == 0x02) {
+            lEsFrame.mPts = lEsFrame.mDts = readPts(rSb);
+        } else if (lPesHeader.mPtsDtsFlags == 0x03) {
+            lEsFrame.mPts = readPts(rSb);
+            // readPts function reads a "timestamp", so we use it for DTS as well

```

Comment:
Why not rename this method to `readTimestamp`?

Reply:
I kept it for now but changed the comment a bit.. (they are "old" functions implemented in common.h/cpp)